azure-pipelines.yml

```diff
@@ -10,24 +10,85 @@ pool:
   vmImage: 'windows-latest'
 
 variables:
-  solution: '**/*.sln'
-  buildPlatform: 'Any CPU'
-  buildConfiguration: 'Release'
+# must be created in the azure UI under Pipelines > Library and specify accessKey, secretKey, region, and bucket
+- group: 's3-public-bucket-upload'
+- name: solution 
+  value: '**/*.sln'
+- name: buildPlatform
+  value: 'Any CPU'
+- name: buildConfiguration
+  value: 'Release'
+- name: installerPath
+  value: "MorphicWin/bin/Release/netcoreapp3.1/Morphic.exe"
 
-steps:
-- task: NuGetToolInstaller@1
+stages:
+- stage: build
+  jobs:
+    - job: build
+      steps:
+        - task: NuGetToolInstaller@1
+        
+        - task: NuGetCommand@2
+          inputs:
+            restoreSolution: '$(solution)'
+        
+        - task: VSBuild@1
+          inputs:
+            solution: '$(solution)'
+            platform: '$(buildPlatform)'
+            configuration: '$(buildConfiguration)'
+        
+        - task: VSTest@2
+          inputs:
+            platform: '$(buildPlatform)'
+            configuration: '$(buildConfiguration)'
 
-- task: NuGetCommand@2
-  inputs:
-    restoreSolution: '$(solution)'
+        - publish: $(System.DefaultWorkingDirectory)/$(installerPath)
+          artifact: installer
 
-- task: VSBuild@1
-  inputs:
-    solution: '$(solution)'
-    platform: '$(buildPlatform)'
-    configuration: '$(buildConfiguration)'
+- stage: upload
+  jobs:
+    - job: publish_client
+      pool:
+        vmImage: "ubuntu-18.04"
+      steps:
+        # do not checkout code. we only need to upload the published artifact
+        - checkout: none
+        - download: current
+          artifact: installer
+        - task: Bash@3
+          env:
+            AWS_ACCESS_KEY_ID: $(accessKey)
+            AWS_SECRET_ACCESS_KEY: $(secretKey)
```

Comment:
Nice. Gets passed in, eh?

Reply:
Yea, it's part of the variable group I import in the "variables" section. Unfortunately that has to be created manually... but it's a one-time thing... so whatever.